### ISSUE TITLE
Existing User Should Not be able to Request Access

### ISSUE THREAD

**Issue by aarontitus** [CrisisCleanup/crisiscleanup-3-web#692]
**Describe the bug**
An existing user is currently allowed to request access to a different organization.

**To Reproduce**
1. Click "Request Access"
2. Complete the form with the email address of an existing user.
3. The request is completed. It should be blocked.

**Expected behavior**
Error message: "It appears you already have an account. If you have forgotten your password, click 'Forgot Password' on the Login page."

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by BradenM** [comment]
Closed by https://github.com/CrisisCleanup/crisiscleanup-3-api/commit/414947c992009a72988bc5fdc622d8aa9c163654

**Comment by aarontitus** [comment]
This is not complete. 
1. Go to https://app.dev.crisiscleanup.io/request_access
2. Target email is admindemo@crisiscleanup.org
3. Email if new user is "Demo@crisiscleanup.org" (note case)
4. The request should be blocked, but is allowed.

Also note: a request to "Admindemo@crisiscleanup.org" (note case) fails, because the search is apparently case sensitive. The search should not be case sensitive. 

When comparing email addresses, make them both lower case, and trim them. Adding a trailing space (which is default behavior for many cell phone keyboards) also leads to a failure.

**Comment by aarontitus** [comment]
One important amendment. An existing user should be able to request access to a DIFFERENT organization. If approved, it is treated as a transfer, which is apparently the existing behavior. I have to double-check this is the case on staging.

**Comment by aarontitus** [comment]
Quick question, are you writing tests for each issue?

**Comment by BradenM** [comment]
Closed by https://github.com/CrisisCleanup/crisiscleanup-3-api/commit/639c37686a4b681fe6d29c6f7284da63df8ca2c7

**Comment by aarontitus** [comment]
This issue is not quite complete, because error handling is incorrect.
1. Go to https://app.dev.crisiscleanup.io/request_access
2. Request access to admindemo@crisiscleanup.org, as demo@crisiscleanup.org. Complete the form.
3. You get a 500 error, "Uh, oh..." This is incorrect. You should get an error, "This email address already has an account. To move organizations, log in and click Profile under your name. Then click the 'Change or Move Organizations'."
4. Now do exactly the opposite. Request access to demo@crisiscleanup.org as admindemo@crisiscleanup.org. Complete the form.
5. Although you _should_ get the same "This email address already has an account..." error, instead you get the error, "This account has an invitation pending."
6. This is wrong for two reasons: 1. A pending invitation should NOT preclude someone from requesting access. 2. The user has an existing account, and should not have any pending invitations.

We should also limit Request Access tries to three before imposing a wait limit of 5 minutes.

**Comment by BradenM** [comment]
@aarontitus Can you try again without using demo user (admin demo is fine) as per our discussion in #709 ?

> Although you should get the same "This email address already has an account..." error, instead you get the error, "This account has an invitation pending."
This is wrong for two reasons: 1. A pending invitation should NOT preclude someone from requesting access. 2. The user has an existing account, and should not have any pending invitations.

The pending invitation was pre-existing behavior. Because I was also testing with admindemo and such, I am going to delete invitations for them so you can test with a clean slate (just in case0. 

So would you like me to remove the invitation pending check? This would effectively allow a user to send multiple requests to the same user/org.

**Comment by aarontitus** [comment]
@BradenM 
1. The system SHOULD check for any existing entries in `user_invitations_requests`. If the person has *requested* an invitation, they should not be allowed to request another invitation. Is that the check that is currently happening? If so, then we should update the error message to read, "You have already requested access to an existing user. They must reject your request before you may make another request." I will update the localizations.
2. An *invitation* should not preclude a new user requesting access to another user. This appears to be the case. No problem.
3. However, if Alice invites Bob, then Carl invites Bob, then Bob requests access from Dan: Bob should have two invitations and one request_invitation. When Bob activates one of the invitations, then the other two invitations/ request_invitations should be marked as rejected by Bob.

Currently, if Alice invites Bob, then Carl invites Bob, Bob only has one invitation in the database marked as "created_by" Alice, but updated_by Carl, and for Carl's organization. This part needs to be fixed.

**Comment by BradenM** [comment]
Ah, gotcha.

Okay, users should now be able to receive multiple invitations, and any outstanding invitations are rejected once they accept one.

[See Video](https://clips.bradenmars.me/2020-05-29T05-05:00-multiuser.mp4)

Note that although it isnt portrayed well, the second invite's `accepted` status is changed from NULL -> False

For request invitations, I updated it to allow a user to make multiple requests as long as they are to different users. If they try to make a request to the same user (before it has been rejected/accepted), they will receive the 'pending invite' error. 

[See Video](https://clips.bradenmars.me/2020-05-29T05-05:00-request_invite.mp4)

If you want to update the "pending invitation" localization, its here:

https://github.com/CrisisCleanup/crisiscleanup-3-api/blob/9f3b93f8e633afc402b42740707ad5373b4586fa/crisiscleanup/users/serializers.py#L523-L527

**Comment by aarontitus** [comment]
@BradenM The only concern I have with allowing users to *request* multiple invitations to the same organization (but different users) is that their names will appear multiple times in the "Current Requests" section of "Invitation Management"
Possible Solutions:
1. Allow users to make multiple requests to different users within the same organization, but only show one in "Invitation Management," then invalidate all of them as soon as one is accepted.
2. Do not allow users to make multiple requests _unless_ the requests are to users from different organizations.
3. Do not allow users to make multiple requests as along as a pending, unresolved _request_access_ exists.

Which do you prefer?

**Comment by BradenM** [comment]
Ah good catch. I will add a check to ensure the users are in different organizations.

**Comment by aarontitus** [comment]
@BradenM This is still not complete.
I attempted to request access to demo@crisiscleanup.org (organization_id = 12) AS admindemo@crisiscleanup.org (organization_id = 89). This results in a success message, but did not result in any pending requests. However, it should have resulted in a failure message for two reasons:
1. admindemo@crisiscleanup.org already has an account. Consequently, this should fail. 
2. After thinking about it further, it does not matter that the request came from a user in different organizations. If the user wants to change organizations, they need to log in and go through the entire process to change organizations. They are not authenticated, so a malicious user could (in theory) move users without permission. Also, the process to move organizations is quite complex and cannot be done here.

**Comment by BradenM** [comment]
Closed by API#311

**Comment by aarontitus** [comment]
@BradenM This is not complete. In the original use case:
1. Click "Request Access"
2. Complete the form with the email address of an existing user.
3. The request is completed. It should be blocked.
Returns the error, "You have already requested access to an existing user. They must reject your request before you may make another request." It should read,
"It appears you already have an account. If you have forgotten your password, click 'Forgot Password' on the Login page."
By the way, I spent WAY too long trying to track down the error message, which I know is not translated. Where is it? Is it on one of your local branches?
![Add additional error](https://user-images.githubusercontent.com/4184619/90471440-87dba580-e0db-11ea-831b-7e7510983ae7.png)

**Comment by BradenM** [comment]
> 3.) The request is completed. It should be blocked.

Unsure what you mean by this. In the image, it appears the request is not being completed and the error message is showing?
Just tried it to double check and it seems to be working fine (that is, NOT creating a new request.)

> Returns the error, "You have already requested access to an existing user. They must reject your request before you may make another request." It should read,
"It appears you already have an account. If you have forgotten your password, click 'Forgot Password' on the Login page."
By the way, I spent WAY too long trying to track down the error message, which I know is not translated. Where is it? Is it on one of your local branches?

:grimacing: Well... it actually **is** translated... and it comes from the Api, which is likely why you couldn't find it.

Here is exactly where the error message is translated:
https://github.com/CrisisCleanup/crisiscleanup-3-api/blob/master/crisiscleanup/users/serializers.py#L628-L634

Translation key: `"info.no_request_access_existing_account"`

Since it seems the localization key was the only issue, I've opened a separate issue here: API#351
